icu4c/source/common/cmemory.h

```diff
@@ -399,8 +405,49 @@ class MaybeStackArray {
     bool operator==(const MaybeStackArray & /*other*/) {return FALSE;}
     bool operator!=(const MaybeStackArray & /*other*/) {return TRUE;}
     /* No ownership transfer: No copy constructor, no assignment operator. */
-    MaybeStackArray(const MaybeStackArray & /*other*/) {}
-    void operator=(const MaybeStackArray & /*other*/) {}
+    MaybeStackArray(const MaybeStackArray & /*other*/) = delete;
+    void operator=(const MaybeStackArray & /*other*/) = delete;
+};
+
+/**
+ * A copyable MaybeStackArray. If memory allocation succeeds, the copy will have
+ * the same capacity as the original, with all values copied (including
+ * uninitialized ones).
+ *
+ * WARNING: does not report errors upon memory allocation failure, after which
+ * capacity will be less than expected!
+ */
+template <typename T, int32_t stackCapacity>
+class CopyableMaybeStackArray : public MaybeStackArray<T, stackCapacity> {
+  public:
+    /**
+     * Default constructor initializes with internal T[stackCapacity] buffer.
+     */
+    CopyableMaybeStackArray() : MaybeStackArray<T, stackCapacity>() {
+    }
+
+    /**
+     * Copy constructor.  If memory allocation succeeds, the copy will have
+     * the same capacity as the original, with all values copied (including
+     * uninitialized ones).
+     *
+     * WARNING: does not report errors upon memory allocation failure, after which
+     * capacity will be less than expected!
+     */
+    CopyableMaybeStackArray(const CopyableMaybeStackArray &other) {
+        this->operator=(other);
+    };
+
+    // Assignment operator
+    void operator=(const CopyableMaybeStackArray &rhs) {
+        int32_t length = rhs.capacity;
+        this->resize(length, length);
+        // If memory allocation failed, capacity will be at least stackCapacity.
+        if (length > this->capacity) {
+            length = this->capacity;
+        }
+        uprv_memcpy(rhs.ptr, this->ptr, (size_t)length * sizeof(T));
+    };
```

Comment:
I think you might have to implement the move constructor and move assignment, or at least explicitly default them, to avoid the copy constructor always being used.

Reply:
Done.

Reply:
(And manually tested/explored with some printf'ing and a debugger, FWIW. I finally think I'm not confused by std::move anymore - it is as straight-forward as I once thought before I got confused. :)